src/engine/cachingreader.cpp

```diff
@@ -205,32 +205,27 @@ void CachingReader::newTrack(TrackPointer pTrack) {
     // to get ready while the reader switches its internal
     // state. There are no race conditions, because the
     // reader polls the worker.
+    m_state.store(State::TrackLoading, std::memory_order_acquire);
     m_worker.newTrack(pTrack);
     m_worker.workReady();
     // Don't accept any new read requests until the current
     // track has been unloaded and the new track has been
     // loaded.
-    m_state = State::TrackLoading;
-    // Free all chunks with sample data from the current track.
-    freeAllChunks();
 }
 
 void CachingReader::process() {
     ReaderStatusUpdate update;
     while (m_stateFIFO.read(&update, 1) == 1) {
-        DEBUG_ASSERT(m_state != State::Idle);
         auto pChunk = update.takeFromWorker();
         if (pChunk) {
+            DEBUG_ASSERT(m_state != State::Idle);
```

Comment:
You don't need to move the debug assertion. The condition must be valid while looping, independent of the update message.

Reply:
At the original position it happen when you have ejected the track.
In this case the "idle" state has overwritten the "loading" state because a null track was loaded. After loading a new non null track we receive the "loaded" signal while in idle state.

Reply:
We should receive TRACK_LOADED/UNLOADED messages only while the state is *Loading*. The *Idle* state is set **after** we have received TRACK_UNLOADED. We should not receive any messages from the worker until the state has been changed from *Idle* to *Loading* again.

Reply:
The issue is that the cue is not processed after ejecting the track. That is why the idle message is still in the queue and overwrite the loading message.
This is not much of an issue, but I see how it is surprising. 
I will look how to solve this.

Reply:
It is not an issue to overwrite the loading message because the very next message is the loaded message and the cue was purged until the loading state was there. Purging is stopped by the idle message. 

Can we solve this by renaming?